PR: https://github.com/odoo/odoo/pull/239963

From: fbbbd3314a1574b1196a59bd4229ee81aa932eb7
From: Jeanne Delneste
Date: 2026-02-13 05:49:25

Breaking data model changes scores: del:5 + add:1, change matches:
-    company_id = fields.Many2one(
-    excluded_journal_ids = fields.Many2many(
+    included_journal_ids = fields.One2many(
-    journal_group_ids = fields.Many2many('account.journal.group',

Total Changes: 63

[IMP] account: inclusion of journals in journal groups

Instead of exclusion of journals, now the journal groups are an inclusion of journals.
The journal group (ledger)  can now be chosen when creating a journal.

task-5404526

closes odoo/odoo#239963

Related: odoo/enterprise#102089
Related: odoo/upgrade#9099
Signed-off-by: William André (wan) <wan@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_journal.py
+++ b/addons/account/models/account_journal.py
@@ -16,26 +16,19 @@ _logger = logging.getLogger(__name__)
 class AccountJournalGroup(models.Model):
     _name = 'account.journal.group'
     _description = "Account Journal Group"
-    _check_company_auto = True
-    _check_company_domain = models.check_company_domain_parent_of
 
     name = fields.Char("Ledger group", required=True, translate=True)
-    company_id = fields.Many2one(
-        comodel_name='res.company',
-        help="Define which company can select the multi-ledger in report filters. If none is provided, available for all companies",
-        default=lambda self: self.env.company,
-    )
-    excluded_journal_ids = fields.Many2many(
-        comodel_name='account.journal',
-        domain='company_id and [("company_id", "parent_of", company_id)] or []',
-        string="Excluded Journals",
-        context={'active_test': False},
+
+    included_journal_ids = fields.One2many(
+        'account.journal',
+        'journal_group_id',
+        string="Included Journals",
     )
     sequence = fields.Integer(default=10)
 
     _uniq_name = models.Constraint(
-        'unique(company_id, name)',
-        'A Ledger group name must be unique per company.',
+        'unique(name)',
+        'A Ledger group name must be unique.',
     )
 
 
@@ -257,9 +250,11 @@ class AccountJournal(models.Model):
                                   "Any file extension will be accepted.\n"
                                   "Only PDF and XML files will be interpreted by Odoo")
 
-    journal_group_ids = fields.Many2many('account.journal.group',
-        check_company=True,
-        string="Ledger Group")
+    journal_group_id = fields.Many2one(
+        comodel_name='account.journal.group',
+        string="Ledger",
+        index='btree_not_null'
+    )
 
     available_payment_method_ids = fields.Many2many(
         comodel_name='account.payment.method',
@@ -295,6 +290,9 @@ class AccountJournal(models.Model):
         'unique (company_id, code)',
         'Journal codes must be unique per company.',
     )
+    has_ledger = fields.Boolean(
+        related='company_id.has_ledger',
+    )
 
     def _compute_has_invalid_statements(self):
         journals_with_invalid_statements = self.env['account.bank.statement'].search([

--- a/addons/account/models/account_move.py
+++ b/addons/account/models/account_move.py
@@ -176,9 +176,9 @@ class AccountMove(models.Model):
     )
     journal_group_id = fields.Many2one(
         'account.journal.group',
+        related='journal_id.journal_group_id',
         string='Ledger',
         store=False,
-        search='_search_journal_group_id',
     )
     company_id = fields.Many2one(
         comodel_name='res.company',
@@ -2518,15 +2518,6 @@ class AccountMove(models.Model):
     # SEARCH METHODS
     # -------------------------------------------------------------------------
 
-    def _search_journal_group_id(self, operator, value):
-        field = 'name' if 'like' in operator else 'id'
-        journal_groups = self.env['account.journal.group'].search([(field, operator, value)])
-        return Domain.OR([
-            Domain('journal_id', 'not in', group.excluded_journal_ids.ids)
-            & Domain('journal_id.company_id', '=?', group.company_id.id)
-            for group in journal_groups
-        ])
-
     def _search_reconciled_payment_ids(self, operator, value):
         if operator not in ('in', '='):
             return NotImplemented

--- a/addons/account/models/account_move_line.py
+++ b/addons/account/models/account_move_line.py
@@ -49,8 +49,8 @@ class AccountMoveLine(models.Model):
     journal_group_id = fields.Many2one(
         string='Ledger',
         comodel_name='account.journal.group',
+        related='journal_id.journal_group_id',
         store=False,
-        search='_search_journal_group_id',
     )
 
     company_id = fields.Many2one(
@@ -1444,13 +1444,6 @@ class AccountMoveLine(models.Model):
             'type': 'ir.actions.act_window',
         }
 
-    # -------------------------------------------------------------------------
-    # SEARCH METHODS
-    # -------------------------------------------------------------------------
-
-    def _search_journal_group_id(self, operator, value):
-        return self.env['account.move']._search_journal_group_id(operator, value)
-
     # -------------------------------------------------------------------------
     # INVERSE METHODS
     # -------------------------------------------------------------------------

--- a/addons/account/models/company.py
+++ b/addons/account/models/company.py
@@ -307,6 +307,11 @@ class ResCompany(models.Model):
         help="During perpetual valuation, this account will hold the price difference between the standard price and the bill price.",
     )
 
+    # If company has Ledgers
+    has_ledger = fields.Boolean(
+        compute='_compute_has_ledger',
+    )
+
     def get_next_batch_payment_communication(self):
         '''
         When in need of a batch payment communication reference (several invoices paid at the same time)
@@ -467,6 +472,12 @@ class ResCompany(models.Model):
     def _compute_qr_code(self):
         pass
 
+    def _compute_has_ledger(self):
+        self.has_ledger = bool(self.env['account.journal.group'].search_count(
+            domain=[],
+            limit=1,
+        ))
+
     def _initiate_account_onboardings(self):
         account_onboarding_routes = [
             'account_dashboard',
